File: server/core/sparkarc_config.py
"""读取仓库根目录的跨语言 SparkArc 常量清单。"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


_PROJECT_ROOT = Path(__file__).resolve().parents[2]
_CONFIG_PATH = _PROJECT_ROOT / "sparkarc.json"
_REQUIRED_NETWORK_RESOURCES = {
    "pypi",
    "npm_registry",
    "github_release",
    "gh_proxy",
    "huggingface",
    "python_standalone",
    "node_distribution",
}


def project_config_path() -> Path:
    """返回受版本控制的跨语言常量文件路径。"""
    return _CONFIG_PATH


def _require_string(value: Any, name: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise RuntimeError(f"sparkarc.json 的 {name} 必须是非空字符串。")
    return value.strip()


def _require_url_list(value: Any, name: str) -> list[str]:
    if not isinstance(value, list) or any(not isinstance(item, str) or not item.strip() for item in value):
        raise RuntimeError(f"sparkarc.json 的 {name} 必须是 URL 字符串数组。")
    return [item.strip() for item in value]


def _validate_config(raw: Any) -> dict[str, Any]:
    if not isinstance(raw, dict) or raw.get("schemaVersion") != 1:
        raise RuntimeError("sparkarc.json 的 schemaVersion 必须为 1。")

    repository = raw.get("repository")
    if not isinstance(repository, dict) or repository.get("provider") != "github":
        raise RuntimeError('sparkarc.json 当前仅支持 repository.provider = "github"。')
    slug = _require_string(repository.get("slug"), "repository.slug")
    if len(slug.split("/")) != 2:
        raise RuntimeError("sparkarc.json 的 repository.slug 必须是 owner/repository 格式。")

    network = raw.get("network")
    if not isinstance(network, dict):
        raise RuntimeError("sparkarc.json 缺少 network 配置。")
    providers = _require_url_list(network.get("geoIpProviders"), "network.geoIpProviders")
    if len(set(providers)) < 2:
        raise RuntimeError("sparkarc.json 的 network.geoIpProviders 至少需要两个不同的服务。")
    resources = network.get("resources")
    if not isinstance(resources, dict):
        raise RuntimeError("sparkarc.json 缺少 network.resources 配置。")
    missing = sorted(_REQUIRED_NETWORK_RESOURCES - set(resources))
    if missing:
        raise RuntimeError(f"sparkarc.json 缺少网络资源：{', '.join(missing)}。")
    for resource_name, route in resources.items():
        if not isinstance(route, dict):
            raise RuntimeError(f"sparkarc.json 的 network.resources.{resource_name} 必须是对象。")
        _require_url_list(route.get("default"), f"network.resources.{resource_name}.default")
        _require_url_list(route.get("mainland"), f"network.resources.{resource_name}.mainland")
    return raw


@lru_cache(maxsize=1)
def load_sparkarc_config() -> dict[str, Any]:
    """加载并校验项目常量；启动期失败优先暴露配置错误。"""
    try:
        raw = json.loads(_CONFIG_PATH.read_text(encoding="utf-8"))
    except OSError as exc:
        raise RuntimeError(f"无法读取 {_CONFIG_PATH}: {exc}") from exc
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"{_CONFIG_PATH} 不是有效 JSON: {exc}") from exc
    return _validate_config(raw)


def repository_urls() -> dict[str, str]:
    """从唯一仓库标识派生公开仓库、克隆与 Release 地址。"""
    slug = str(load_sparkarc_config()["repository"]["slug"])
    web = f"https://github.com/{slug}"
    return {
        "slug": slug,
        "web": web,
        "clone": f"{web}.git",
        "release_api": f"https://api.github.com/repos/{slug}/releases/latest",
        "release_page": f"{web}/releases/latest",
    }


def network_candidates(
    resource: str,
    *,
    mainland: bool,
    include_fallback: bool = True,
) -> list[str]:
    """按出口区域返回资源候选，保持配置中的优先级。"""
    resources = load_sparkarc_config()["network"]["resources"]
    route = resources.get(resource)
    if not isinstance(route, dict):
        raise ValueError(f"sparkarc.json 未定义网络资源：{resource}")
    preferred_key, fallback_key = ("mainland", "default") if mainland else ("default", "mainland")
    values = list(route[preferred_key])
    if include_fallback:
        values.extend(route[fallback_key])
    return list(dict.fromkeys(str(value).strip() for value in values if str(value).strip()))


def geoip_providers() -> list[str]:
    """返回用于无代理直连投票的出口地址查询服务。"""
    return list(load_sparkarc_config()["network"]["geoIpProviders"])

File: scripts/sparkarc-config.mjs
/**
 * SparkArc 跨语言项目常量的 Node 读取器。
 *
 * 业务代码不应在此处维护备用网址或仓库地址；所有值来自根目录 sparkarc.json。
 */
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const scriptDir = dirname(fileURLToPath(import.meta.url));
export const projectRoot = join(scriptDir, '..');
export const configPath = join(projectRoot, 'sparkarc.json');

function requireString(value, name) {
  if (typeof value !== 'string' || !value.trim()) {
    throw new Error(`sparkarc.json 的 ${name} 必须是非空字符串。`);
  }
  return value.trim();
}

function requireUrlList(value, name) {
  if (!Array.isArray(value) || value.some((item) => typeof item !== 'string' || !item.trim())) {
    throw new Error(`sparkarc.json 的 ${name} 必须是 URL 字符串数组。`);
  }
  return value.map((item) => item.trim());
}

export function validateSparkArcConfig(config) {
  if (!config || typeof config !== 'object' || config.schemaVersion !== 1) {
    throw new Error('sparkarc.json 的 schemaVersion 必须为 1。');
  }
  if (config.repository?.provider !== 'github') {
    throw new Error('sparkarc.json 当前仅支持 repository.provider = "github"。');
  }
  const slug = requireString(config.repository?.slug, 'repository.slug');
  if (!/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/.test(slug)) {
    throw new Error('sparkarc.json 的 repository.slug 必须是 owner/repository 格式。');
  }
  if (config.repository?.mainlandRelease?.provider !== 'gitee') {
    throw new Error('sparkarc.json 当前仅支持 repository.mainlandRelease.provider = "gitee"。');
  }
  const mainlandReleaseSlug = requireString(config.repository.mainlandRelease.slug, 'repository.mainlandRelease.slug');
  if (!/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/.test(mainlandReleaseSlug)) {
    throw new Error('sparkarc.json 的 repository.mainlandRelease.slug 必须是 owner/repository 格式。');
  }
  requireUrlList(config.repository?.mainlandCloneUrls, 'repository.mainlandCloneUrls');
  if (!config.network || typeof config.network !== 'object') {
    throw new Error('sparkarc.json 缺少 network 配置。');
  }
  const geoIpProviders = requireUrlList(config.network.geoIpProviders, 'network.geoIpProviders');
  if (new Set(geoIpProviders).size < 2) {
    throw new Error('sparkarc.json 的 network.geoIpProviders 至少需要两个不同的服务。');
  }
  if (!config.network.resources || typeof config.network.resources !== 'object') {
    throw new Error('sparkarc.json 缺少 network.resources 配置。');
  }
  for (const [name, route] of Object.entries(config.network.resources)) {
    requireUrlList(route?.default, `network.resources.${name}.default`);
    requireUrlList(route?.mainland, `network.resources.${name}.mainland`);
  }
  for (const requiredResource of ['pypi', 'npm_registry', 'github_release', 'gh_proxy', 'huggingface', 'python_standalone', 'node_distribution']) {
    if (!Object.hasOwn(config.network.resources, requiredResource)) {
      throw new Error(`sparkarc.json 缺少 network.resources.${requiredResource}。`);
    }
  }
  return config;
}

export function readSparkArcConfig() {
  let config;
  try {
    config = JSON.parse(readFileSync(configPath, 'utf8'));
  } catch (error) {
    throw new Error(`无法读取 ${configPath}: ${error instanceof Error ? error.message : String(error)}`);
  }
  return validateSparkArcConfig(config);
}

export function repositoryUrls(config = readSparkArcConfig()) {
  const slug = config.repository.slug;
  const web = `https://github.com/${slug}`;
  const mainlandReleaseSlug = config.repository.mainlandRelease.slug;
  const mainlandReleaseWeb = `https://gitee.com/${mainlandReleaseSlug}`;
  return {
    slug,
    web,
    clone: `${web}.git`,
    mainlandClones: [...config.repository.mainlandCloneUrls],
    releaseApi: `https://api.github.com/repos/${slug}/releases/latest`,
    releasePage: `${web}/releases/latest`,
    mainlandRelease: {
      provider: config.repository.mainlandRelease.provider,
      slug: mainlandReleaseSlug,
      web: mainlandReleaseWeb,
      releaseApi: `https://gitee.com/api/v5/repos/${mainlandReleaseSlug}/releases/latest`,
      releasePage: `${mainlandReleaseWeb}/releases/latest`,
    },
  };
}

export function networkCandidates(resource, { mainland = false, includeFallback = true, config = readSparkArcConfig() } = {}) {
  const route = config.network.resources[resource];
  if (!route) {
    throw new Error(`sparkarc.json 未定义网络资源 ${resource}。`);
  }
  const preferred = mainland ? route.mainland : route.default;
  const fallback = mainland ? route.default : route.mainland;
  const candidates = includeFallback ? [...preferred, ...fallback] : [...preferred];
  return [...new Set(candidates.map((item) => item.trim()).filter(Boolean))];
}

export function allNetworkCandidates(resource, config = readSparkArcConfig()) {
  const route = config.network.resources[resource];
  if (!route) {
    throw new Error(`sparkarc.json 未定义网络资源 ${resource}。`);
  }
  return [...new Set([...route.default, ...route.mainland].map((item) => item.trim()).filter(Boolean))];
}
